app/src/main/java/com/guzelgimadieva/tasky/core/network/TaskyServiceImpl.kt

```diff
@@ -0,0 +1,67 @@
+package com.guzelgimadieva.tasky.core.network
+import com.guzelgimadieva.tasky.core.data.remote.model.LoginRequest
+import com.guzelgimadieva.tasky.core.data.remote.model.LoginResponse
+import com.guzelgimadieva.tasky.core.data.remote.model.RegisterRequest
+import com.guzelgimadieva.tasky.BuildConfig
+import io.ktor.client.HttpClient
+import io.ktor.client.call.body
+import io.ktor.client.engine.okhttp.OkHttp
+import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.client.plugins.defaultRequest
+import io.ktor.client.plugins.logging.Logger
+import io.ktor.client.plugins.logging.Logging
+import io.ktor.client.plugins.logging.SIMPLE
+import io.ktor.client.request.header
+import io.ktor.client.request.post
+import io.ktor.client.request.setBody
+import io.ktor.client.statement.HttpResponse
+import io.ktor.http.ContentType
+import io.ktor.http.contentType
+import io.ktor.http.headers
+import io.ktor.serialization.kotlinx.json.json
+import kotlinx.serialization.json.Json
+
+class TaskyServiceImpl: TaskyService {
+
+    private val client = HttpClient(OkHttp){
+        install(Logging) {
+            logger = Logger.SIMPLE
+        }
+        defaultRequest {
+            header("x-api-key", BuildConfig.API_KEY)
+            contentType(ContentType.Application.Json)
+        }
+        install(ContentNegotiation) {
+            json(
+                Json {
+                    isLenient = true
+                    ignoreUnknownKeys = true
+                }
+            )
+        }
+
+    }
+    override suspend fun register(registerRequest: RegisterRequest) {
+        safeRequest {
+            client.post(HttpRoutes.REGISTER) {
+                setBody(registerRequest)
+            }
+        }
+    }
+
+    override suspend fun login(loginRequest: LoginRequest): NetworkResponseState<LoginResponse> {
+       return safeRequest {
+            client.post(HttpRoutes.LOGIN){
+                setBody(loginRequest)
+            }.body<LoginResponse>()
+        }
+    }
+
+    private inline fun <T> safeRequest(request: () -> T): NetworkResponseState<T> {
+        return try {
+            NetworkResponseState.Success(data = request())
+        } catch (e: Exception) {
+            NetworkResponseState.Error(exception = e)
+        }
```

Comment:
Be careful with catching general exceptions inside suspend functions. You can break coroutine cancellation that way. If a coroutine is cancelled, the currently suspending function (here, it would be `FUNCTION_NAME`) will throw a CancellationException. If your catch block eats that up, this exception won't be properly propagated up the coroutine scopes which can cause some coroutines to run longer than they should.
You can fix this by adding this:

```
if(e is CancellationException) throw e
```
or alternatively by only catching specific exceptions and not CancellationExceptions.

In this video I show the issue with that a bit more in detail (starting at 17:45):
https://www.youtube.com/watch?v=VWlwkqmTLHc